### DISCUSSION TITLE
Question regarding design philosophy with MVVM

### DISCUSSION THREAD

**Discussion by kuiperzone** [AvaloniaUI/Avalonia#7546] [Q&A] [answered]
I started out writing a code-behind app with Avalonia. As time went on, it become necessary to introduce ViewModels so I have a code-behind app with ViewModels for certain things, such as theme sensitive icons. Overtime, I've been shifting much of the functionality from the Window classes into the ViewModels which got me thinking.

I'm starting to think that, if you're going to use MVVM, then it's best to put almost all functionality into ViewModels and have the Window/Control classes as a light-weight "container" with little functionality beyond creation of the DataContext and the handling of the few things which can't be done in a model. Certainly, it doesn't make sense in my mind to have functionality split in a arbitrary fashion between code-behind and models.

Is this correct? Any thoughts?

**Comment by wieslawsoltes** [comment]
You can put a lot of code into attached behaviors that can be reused across controls.

https://github.com/wieslawsoltes/AvaloniaBehaviors
https://github.com/wieslawsoltes/AvaloniaBehaviors/tree/master/samples

**Comment by maxkatz6** [comment]
> and have the Window/Control classes as a light-weight "container" with little functionality beyond creation of the DataContext and the handling of the few things which can't be done in a model.

It’s easier to explain if you think about these layers in that way:
Model layer - business logic (domain models, services, data access possibly). This layer usually heavily unit tested as well, if you wish. 
View later - UI logic (controls, templates, styles).
ViewModel - glue between two from above.

Now, if you ask what’s allowed to be inside of your window’s code behind - it’s UI logic. You can handle some events in a special way and call view model to execute command for example. Mvvm pattern isn’t about “only xaml code allowed in view”. But there are projects where this strictness is preferred. And tools like Behaviors SDK really help with it.

**Reply by maxkatz6** [reply]
Additional reasoning to avoid code in code behind is about reusing (of behaviors, converters or attached properties), keeping related logic together in one language-layer (xaml) or even delegating view layer to a designer, who doesn’t know c# (rare).

**Reply by maxkatz6** [reply]
Remember that patterns aren’t mandatory. They are created to share common architectures among developers, so it’s easier to understand a new app. And they usually help with some best practices, like unit testing. Just imagine how hard it would be to test some business logic, like tax calculation, if it heavily depends on some text boxes and buttons instead of input and output parameters of simple method.

**Comment by kuiperzone** [comment]
Thank you.

To elaborate a little, the situation I faced was moving from (say Click) events which called on a Windows class to respond, to Commands which called on the DataContext (ViewModel) instance. But what should the ViewModel instance do when called? It could a. Invoke an event on the Windows class (but this goes round the houses and seems non-sensical to me), b. Have access to the Window class reference and call a method on it (which is what I've done), or c. Have everything it needs to do the operation, such as opening other windows etc. (not sure this goes against or is with what you've just said).

**Reply by robloo** [reply]
> what should the ViewModel instance do when called? 

The ViewModel instance, when receiving a command, should update the data structures internal to it. Those data structures are bound to the View (UI Xaml layer). The data binding will take care of the View updates and update all individual controls. Therefore, you really shouldn't be manually calling back into the View/Window code in pure MVVM.

> a. Invoke an event on the Windows class (but this goes round the houses and seems non-sensical to me),
> b. Have access to the Window class reference and call a method on it (which is what I've done),

This doesn't usually make sense. The ViewModel should be using data binding to send updates to the View (Window class). Note that this is somewhat dependent on what you are doing exactly. If you are opening a message dialog or something it's a little more tricky but Avalonia has techniques for that in a pure MVVM app (I'm not familiar with them though).

> c. Have everything it needs to do the operation, such as opening other windows etc. (not sure this goes against or is with what you've just said).

Yes, the ViewModel should have everything it needs to do the operation. It contains ALL data in the view so all information is available to it when running any code logic. Again, view updates generally should be triggered by data binding.

**Comment by robloo** [comment, marked as answer]
> Is this correct? Any thoughts?

Yes, this is correct. It's kind-of how MVVM came about as it just made sense to extract traditional code-behind into base classes as another layer. Then with data-binding/commands Views became very light-weight on purpose. Avalonia takes this a step further and seems to prefer pure MVVM in several places. For example, a lot of events are missing from TextBox as it's expected to just use binding and property changes.

Pure MVVM is also great when targeting multiple XAML UI frameworks. It would be very easy to use the same ViewModel with a WPF and Avalonia app for example and then just wire up bindings in XAML. It's quite clean without additional conditional code. That said, it's usually not worth the trouble IMO to be extremely strict. It introduces a lot of potentially duplicate code where a property from the Model is copied to the ViewModel, etc. This doesn't usually make sense for small or specialized LOB applications. However, for large applications these MVVM abstractions and even property duplication actually help a lot. It defines clear API/Interface surface area between parts of the code base which helps when many people are working on the same code base.

Overall though, classical VB/WinForms-style programming with code-behind and events invoking code that drives view updates does still make sense. UI frameworks should not force applications one way or another and its really up to the application to make such architecture decisions, even on a view-by-view basis.

**Comment by kuiperzone** [comment]
That's great. Makes sense to me now. Thank you all for responding in depth.
